### ISSUE TITLE
System.IO.FileSystem.AccessControl.Tests fails 14 tests on my computer

### ISSUE THREAD

**Issue by AlexRadch** [dotnet/corefx#27629]
System.IO.FileSystem.AccessControl.Tests fails 14 tests on my computer with next log:
```
  Starting:    System.IO.FileSystem.AccessControl.Tests
     System.IO.FileSystemSecurityTests.RemoveAccessRule_IdenticalRule_Succeeds [FAIL]
        Assert.Equal() Failure
        Expected: 0
        Actual:   1
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\FileSystemSecurityTests.cs(161,0): at System.IO.FileSystemSecurityTests.RemoveAccessRule_IdenticalRule_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAuditRuleSpecific_NoMatchableRuleFound [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(195,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAuditRuleSpecific_NoMatchableRuleFound()
     System.IO.FileSystemSecurityTests.RemoveAccessRuleSpecific_Succeeds [FAIL]
        Assert.Equal() Failure
        Expected: 0
        Actual:   1
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\FileSystemSecurityTests.cs(223,0): at System.IO.FileSystemSecurityTests.RemoveAccessRuleSpecific_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.SetAccessRule_AccessControlType_Deny_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(692,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.SetAccessRule_AccessControlType_Deny_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAuditRule_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(232,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAuditRule_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.SetAuditRule_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(275,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.SetAuditRule_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAccessRuleSpecific_AccessControlType_Deny_NoMatchableRules_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(455,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAccessRuleSpecific_AccessControlType_Deny_NoMatchableRules_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.ResetAccessRule_AccessControlType_Deny_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(624,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.ResetAccessRule_AccessControlType_Deny_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.AddAccessRule_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(728,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.AddAccessRule_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAccessRule_AccessControlType_Deny_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(388,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveAccessRule_AccessControlType_Deny_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.AddAuditRule_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(311,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.AddAuditRule_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveRule_AccessControlType_Allow_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(351,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.RemoveRule_AccessControlType_Allow_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.SetAccessRule_AccessControlType_Allow_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(661,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.SetAccessRule_AccessControlType_Allow_Succeeds()
     System.Security.AccessControl.DirectoryObjectSecurityTests.ResetAccessRule_AccessControlType_Allow_Succeeds [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
           C:\Code\corefx\src\System.IO.FileSystem.AccessControl\tests\DirectoryObjectSecurityTests.cs(587,0): at System.Security.AccessControl.DirectoryObjectSecurityTests.ResetAccessRule_AccessControlType_Allow_Succeeds()
  Finished:    System.IO.FileSystem.AccessControl.Tests
```

**Comment by karelz** [comment]
@AlexRadch can you please debug the failures? I expect they are failing only on your system / some sepcial configs, so we will need more information from you to make the issue actionable. Thanks!

**Comment by AlexRadch** [comment]
@karelz may be my windows does not support something with audit?

**Comment by AlexRadch** [comment]
The issues are with localization of `IdentityReference`. For example on Russian Windows to get `WellKnownSidType.LocalSystemSid` you can use english string @"NT AUTHORITY\SYSTEM" or localized string @"NT AUTHORITY\СИСТЕМА".

I think `class IdentityReference` use string comparison for `operator ==` but in localized Windows the same IdentityReference can have two names. English name and localized Name.

So I thing `class IdentityReference` should implement more smart comparison for `operator ==`.

**Comment by AlexRadch** [comment]
I think this is not 'test-bug'.

**Comment by karelz** [comment]
I let @JeremyKuhne @pjanotti decide.

**Comment by JeremyKuhne** [comment]
@bartonjs Can you comment on the `IdentityReference` comments @AlexRadch brought up?

**Comment by bartonjs** [comment]
https://referencesource.microsoft.com/#mscorlib/system/security/principal/ntaccount.cs,222 and http://source.dot.net/#System.Security.Principal.Windows/System/Security/Principal/NTAccount.cs,199 agree that it's just a string comparison.  I don't know enough about the real-world usage to know if it's a good idea to change that or not.

**Comment by JeremyKuhne** [comment]
Presumably using `new SecurityIdentifier(WellKnownSidType.LocalSystemSid, null)` in the tests would get the proper localized behavior? Can you validate @AlexRadch?

> I thing class IdentityReference should implement more smart comparison for operator ==.

Probably not doable as this behavior has always been this way. Adding new behavior to `IdentifyReference` or `GetAccessRules` might be possible with an overload, but changing the existing behavior of the current APIs is likely to wreak havok on existing code. (And would have a performance impact to look up the Sid.)

It isn't a great idea to use `NTAccount` for the built-in accounts (due to what you're seeing). I don't see clear guidance on this, we could definitely update the docs to discourage/encourage better usage patterns.

In the meantime, we have a few separate issues here:

1. The tests aren't internationalization friendly
2. We might want to investigate new APIs that encourage better localized usage
3. We should make sure our docs are clearly giving appropriate guidance

We can't make any API additions in 2.1 and even if we decided to change the behavior of the existing APIs (which is unlikely) we wouldn't be able to make a change of that scope at this point in the cycle. As such I'm moving this out of the 2.1 milestone.

**Comment by bartonjs** [comment]
SID-equality should be more stable, and should be a reasonable change to the tests (to fix the test i18n problem)

**Comment by AlexRadch** [comment]
> SID-equality should be more stable, and should be a reasonable change to the tests (to fix the test i18n problem)

I created issue because that is not only tests issue.

**Comment by AlexRadch** [comment]
> Presumably using new SecurityIdentifier(WellKnownSidType.LocalSystemSid, null) in the tests would get the proper localized behavior?

I think it would. I will created such PR later.

**Comment by JeremyKuhne** [comment]
> I think it would. I will created such PR later.

Thanks!

> I created issue because that is not only tests issue.

I assume the only issue is what I've described (that you can't reliably compare `NTAuthority` directly).  If I'm missing something please let me know.

**Comment by JeremyKuhne** [comment]
Fixed with #28405.
